[TensorExpr fuser] Guard nodes that have tensor output properties determined by non-tensor inputs

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44138 [JIT] Remove references to no longer generated _tanh_backward and _sigmoid_backward
* **#44137 [TensorExpr fuser] Guard nodes that have tensor output properties determined by non-tensor inputs**

We only insert guards on Tensor types, so we rely on the output
of a node being uniquely determined by its input types.
bail if any non-Tensor input affects the output type
and cannot be reasoned about statically

Differential Revision: [D23543602](https://our.internmc.facebook.com/intern/diff/D23543602)

Commit:
Guard nodes that have tensor output properties determined by non-tensor inputs

[ghstack-poisoned]

ZolotukhinM commented on test/jit/test_profiler.py:
```diff
@@ -31,6 +31,29 @@ def tearDown(self):
+        scalar_type_input(x, x, torch.tensor(1))
+        scalar_type_input(x, x, torch.tensor(1.0))
+        g = torch.jit.last_executed_optimized_graph()
+        FileCheck().check("TensorExpr").check("Scalar = aten::item").check_next("Tensor = aten::add").check("TensorExpr").run(g)
```
Nit: might be useful to have a comment describing what IR we're expecting to get and why. Also, the test could be made more targeted if written as in `test_te_fuser_pass.cpp` where the fuser pass is invoked explicitly.

ZolotukhinM commented on test/jit/test_profiler.py:
```diff
@@ -31,6 +31,29 @@ def tearDown(self):
+        non_const_dtype(x, x, True)
+        non_const_dtype(x, x, True)
+        g = torch.jit.last_executed_optimized_graph()
+        FileCheck().check("TensorExpr").check("TensorExpr").check_not("aten::sum")
```
Ditto.

ZolotukhinM reviewed (approved):
👍

Commit:
Update on "[TensorExpr fuser] Guard nodes that have tensor output properties determined by non-tensor inputs"


We only insert guards on Tensor types, so we rely on the output
of a node being uniquely determined by its input types.
bail if any non-Tensor input affects the output type
and cannot be reasoned about statically

[ghstack-poisoned]
